src/main/kotlin/org/jetbrains/research/testspark/helpers/psi/kotlin/KotlinPsiHelper.kt

```diff
@@ -0,0 +1,200 @@
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+            currentLevelClasses = mutableListOf<PsiClassWrapper>().apply { addAll(tempListOfClasses) }
+            interestingPsiClasses.addAll(tempListOfClasses)
+        }
+        log.info("There are ${interestingPsiClasses.size} interesting psi classes")
+        return interestingPsiClasses.toMutableSet()
+    }
+
+    override fun getInterestingPsiClassesWithQualifiedNames(
+        cut: PsiClassWrapper,
+        psiMethod: PsiMethodWrapper,
+    ): MutableSet<PsiClassWrapper> {
+        val interestingPsiClasses = cut.getInterestingPsiClassesWithQualifiedNames(psiMethod)
+        log.info("There are ${interestingPsiClasses.size} interesting psi classes from method ${psiMethod.methodDescriptor}")
+        return interestingPsiClasses
+    }
+
+    override fun getCurrentListOfCodeTypes(e: AnActionEvent): Array<*>? {
+        val result: ArrayList<String> = arrayListOf()
+        val caret: Caret =
+            e.dataContext.getData(CommonDataKeys.CARET)?.caretModel?.primaryCaret ?: return result.toArray()
+
+        val ktClass = getSurroundingClass(caret.offset)
+        val ktFunction = getSurroundingMethod(caret.offset)
+        val line: Int? = getSurroundingLine(caret.offset)?.plus(1)
+
+        ktClass?.let { result.add(getClassDisplayName(it)) }
+        ktFunction?.let { result.add(getMethodDisplayName(it)) }
+        line?.let { result.add(getLineDisplayName(it)) }
+
+        if (ktClass != null && ktFunction != null) {
+            log.info(
+                "The test can be generated for: \n " +
+                    " 1) Class ${ktClass.qualifiedName} \n" +
+                    " 2) Method ${ktFunction.name} \n" +
+                    " 3) Line $line",
+            )
+        }
+
+        return result.toArray()
+    }
+
+    override fun getLineDisplayName(line: Int) = "<html><b><font color='orange'>line</font> $line</b></html>"
+
+    override fun getClassDisplayName(psiClass: PsiClassWrapper): String =
+        "<html><b><font color='orange'>${psiClass.classType.representation}</font> ${psiClass.qualifiedName}</b></html>"
+
+    override fun getMethodDisplayName(psiMethod: PsiMethodWrapper): String {
+        psiMethod as KotlinPsiMethodWrapper
+        return when {
+            psiMethod.isTopLevelFunction -> "<html><b><font color='orange'>top-level function</font></b></html>"
+            psiMethod.isSecondaryConstructor -> "<html><b><font color='orange'>secondary constructor</font></b></html>"
+            psiMethod.isPrimaryConstructor -> "<html><b><font color='orange'>constructor</font></b></html>"
+            psiMethod.isDefaultMethod -> "<html><b><font color='orange'>default method</font> ${psiMethod.name}</b></html>"
+            else -> "<html><b><font color='orange'>method</font> ${psiMethod.name}</b></html>"
+        }
+    }
```

Comment:
Can it be re-written with `return when(psiMethod as KotlinPsiMethodWrapper) { ... }`? 

If possible, we can omit `psiMethod.[something]` in the cases below, which, in my opinion, might be a bit better.

Reply:
If I am not mistaken, this is an invalid syntax